test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/ProjectSystems/LegacyPackageReferenceProjectTests.cs

```diff
@@ -1129,6 +1129,170 @@ public async Task GetInstalledVersion_WithoutPackages_WithAssets_ReturnsEmpty()
+                var cache_packages = await testProject.GetAllPackagesAsync(CancellationToken.None);
+
+                // Assert
+                packages.Item1.Should().Contain(a => a.PackageIdentity.Equals(new PackageIdentity("packageA", new NuGetVersion("2.15.3"))));
+                packages.Item2.Should().Contain(a => a.PackageIdentity.Equals(new PackageIdentity("packageB", new NuGetVersion("1.0.0"))));
+                packages.Item2.Should().Contain(a => a.PackageIdentity.Equals(new PackageIdentity("packageC", new NuGetVersion("2.1.43"))));
+
+                cache_packages.Item1.Should().Contain(a => a.PackageIdentity.Equals(new PackageIdentity("packageA", new NuGetVersion("2.15.3"))));
+                cache_packages.Item2.Should().Contain(a => a.PackageIdentity.Equals(new PackageIdentity("packageB", new NuGetVersion("1.0.0"))));
+                cache_packages.Item2.Should().Contain(a => a.PackageIdentity.Equals(new PackageIdentity("packageC", new NuGetVersion("2.1.43"))));
+            }
+        }
+
+        [Fact]
+        public async Task GetTransitivePackagesAsync_WithNoTransitivePackageReferences_ReturnsOnlyInstalledPackageIdentities()
+        {
+            using (var testDirectory = TestDirectory.Create())
+            {
+                // Setup
+                LegacyPackageReferenceProject testProject = CreateLegacyPackageReferenceProject(testDirectory, "[1.0.0, )");
+
+                var settings = NullSettings.Instance;
+                var context = new DependencyGraphCacheContext(NullLogger.Instance, settings);
+
+                var packageSpecs = await testProject.GetPackageSpecsAsync(context);
+
+                // Package directories
+                var sources = new List<PackageSource>();
+                var packagesDir = new DirectoryInfo(Path.Combine(testDirectory, "globalPackages"));
+                var packageSource = new DirectoryInfo(Path.Combine(testDirectory, "packageSource"));
+                packagesDir.Create();
+                packageSource.Create();
+                sources.Add(new PackageSource(packageSource.FullName));
+
+                var logger = new TestLogger();
+                var request = new TestRestoreRequest(packageSpecs[0], sources, packagesDir.FullName, logger)
+                {
+                    LockFilePath = Path.Combine(testDirectory, "obj", "project.assets.json")
+                };
+
+                await SimpleTestPackageUtility.CreateFullPackageAsync(packageSource.FullName, "packageA", "2.15.3");
+
+                // Act
+                var command = new RestoreCommand(request);
+                var result = await command.ExecuteAsync();
+                await result.CommitAsync(logger, CancellationToken.None);
+                Assert.True(result.Success);
+                var packages = await testProject.GetAllPackagesAsync(CancellationToken.None);
+                var cache_packages = await testProject.GetAllPackagesAsync(CancellationToken.None);
```

Comment:
This test doesn't do anything to validate that the packages were actually cached. So if there was a bug and the assets file is re-read every time, this test wouldn't fail. Hence, calling `GetAllPackagesAsync` and asseting the contents doesn't actually prevent any regressions.

Reply:
Added a check that the assets file was not accessed when we were expecting the packages to come from the cache.

Reply:
I can't find information on Microsoft's own websites, but [this post](https://www.sevenforums.com/tutorials/243272-last-access-timestamp-enable-disable-windows.html) says that Windows Vista, 7 and 8 all have updates to file last access attribute disabled. [This post](https://www.tenforums.com/tutorials/139015-enable-disable-ntfs-last-access-time-stamp-updates-windows-10-a.html) says that in Windows 10 it's enabled for "volumes" (partitions?) smaller than 128GB and disabled if larger.

I can't find information about whether it's enabled or disabled by default on Windows Server 2019, but our CI agents are running Windows Server 2019. From memory, I've read emails in the microbuild list saying that the disks are 500GB, so if Windows Server 2019 follows Windows 10's defaults based on volume size, it will be disabled.

Therefore, unfortunately this approach will lead to false negatives if there's a bug which breaks caching.

Also, from a "single responsibility" point of view, none of the test names say anything about caching. If we have a test to ensure caching works, then we could have just one test, which also reduces business logic duplication, rather than duplicating the test assertions 6 times in other tests.

Reply:
I changed this to delete the assets file before getting the packages from the cache, then check that we get the correct packages and the assets file has not been recreated.
Also moved the cache test into a separate test.

Reply:
github didn't let me comment on this thread as part of the review I just posted.

I think deleting the assets file in the test is testing the implementation, not testing the expected behavior.

If the assets file was deleted, I'd expect the API to stop returning transitive packages, since we no longer know the state of the project, and whether those transitive packages are correct any longer. If this scenario does turn out to be a problem in the future, and the bug fixed, then this test will return a false negative, and the person who fixes the bug is going to have to spend time to understand the test and realize the test needs to be changed.

So, I think it would be better if the test changed the contents of the file to something that will not parse as json, but keep the same file timestamp, which is exactly the "business logic" that the caching code uses to determine if the assets file is up to date or not.

Having said that, it's hard to come up with a scenario where this will happen (someone does git clean on the command line, then checks out a different branch, causing the project to have different packages, and something happens in VS that triggers this NuGet API call before a restore happens, causing the wrong packages to be returned), and hard to reason about whether it will REALLY cause a problem or not, so I'm not going to block the PR on this. But it doesn't feel "right" to me.

Reply:
I changed this to modify the file instead of deleting it.